File: 프로그래머스/2/42626. 더 맵게/더 맵게.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

int solution(vector<int> scoville, int K) {
    int answer = 0;
    priority_queue<int, vector<int>, greater<int>> pq;
    int cnt = 0;
    bool b = false;
    
    for(auto a : scoville){
        pq.push(a);
    }
    
    while(pq.size() >= 2){
        
        if(pq.top() >= K){
            b = true;
            break;
        }
        
        cnt++;
        int a = pq.top();
        pq.pop();
        int b = pq.top();
        pq.pop();
        
        pq.push(a + 2*b);
    }
    
    
    if(pq.top() >= K)
        b = true;
    
    answer = b ? cnt : -1;
    return answer;
}

File: 프로그래머스/2/42626. 더 맵게/더 맵게.java
import java.util.*;

class Solution {
    
    public static int solution(int[] scoville, int K) {
        
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        int cnt = 0;
        boolean flag = false;

        for (int s : scoville) pq.offer(s);
        

        while (pq.size() >= 2) {
            
            if (pq.peek() >= K) {
                flag = true;
                break;
            }

            int a = pq.poll();
            int b = pq.poll();

            pq.offer(a + 2 * b);
            cnt++;
        }

        if (!pq.isEmpty() && pq.peek() >= K) flag = true;
        

        return flag ? cnt : -1;
    }
}